File: app/src/main/java/me/tatarka/betterfragment/sample/dagger/DaggerFragmentFactory.java
package me.tatarka.betterfragment.sample.dagger;

import java.util.Map;

import javax.inject.Inject;
import javax.inject.Provider;
import javax.inject.Singleton;

import androidx.annotation.NonNull;
import me.tatarka.betterfragment.Fragment;

@Singleton
public class DaggerFragmentFactory implements Fragment.Factory {

    private final Map<Class<? extends Fragment>, Provider<Fragment>> fragmentMap;

    @Inject
    public DaggerFragmentFactory(Map<Class<? extends Fragment>, Provider<Fragment>> fragmentMap) {
        this.fragmentMap = fragmentMap;
    }

    @NonNull
    @Override
    @SuppressWarnings("unchecked")
    public <T extends Fragment> T newInstance(@NonNull Class<T> fragmentClass) {
        return (T) fragmentMap.get(fragmentClass).get();
    }
}

File: app/src/main/java/me/tatarka/betterfragment/sample/dagger/DaggerFragmentFactory.kt
package me.tatarka.betterfragment.sample.dagger

import android.os.Bundle
import me.tatarka.betterfragment.app.Fragment
import javax.inject.Inject
import javax.inject.Provider
import javax.inject.Singleton

@Singleton
class DaggerFragmentFactory @Inject constructor(
    fragmentMap: @JvmSuppressWildcards Map<Class<out Fragment>, Provider<Fragment>>
) : Fragment.Factory {

    private val fragmentMap = fragmentMap.mapKeys { it.key.name }

    override fun <T : Fragment> newInstance(name: String, args: Bundle): T =
        fragmentMap.getValue(name).get() as T
}
